iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,380 @@
+                    let (cid, bytes) = TreeNode::Stem(links).encode()?;
+                    let len = bytes.len();
+                    yield (cid, bytes);
+
+                    // add cid to parent node
+                    tree[i+1].push((cid, len));
+                }
+                // at this point the tree will be able to recieve new links
+                // without "overflowing", aka the leaf node and stem nodes
+                // have fewer than max_degrees number of links
+            }
+
+            // now that we know the tree is in a "healthy" state to
+            // recieve more links, add the link to the tree
+            let (cid, bytes) = TreeNode::Leaf(chunk).encode()?;
+            let len = bytes.len();
+            tree[0].push((cid, len));
+            yield (cid, bytes);
+            // at this point, the leaf node may have max_degrees number of
+            // links, but no other stem node will
+        }
+
+        // our stream had 1 chunk that we have already yielded
+        if tree.len() == 1 && tree[0].len() == 1 {
+            return
+        }
+
+        // clean up, aka yield the rest of the stem nodes
+        // since all the stem nodes are able to recieve links
+        // we don't have to worry about "overflow"
+        while let Some(links) = tree.pop_front() {
+            // root node
+            // if tree.len() == 0 {
+            //     let (cid, bytes) = TreeNode::Root(links).encode()?;
+            //     yield (cid, bytes);
+            //     return
+            // }
+
+            let (cid, bytes) = TreeNode::Stem(links).encode()?;
+            let len = bytes.len();
+            yield (cid, bytes);
+
+            if let Some(front) = tree.front_mut() {
+                front.push((cid, len));
+            } else {
+                // final root, nothing to do
+            }
+        }
+    }
+}
+
+fn create_unixfs_node_from_links(links: Vec<(Cid, usize)>, is_root: bool) -> Result<UnixfsNode> {
+    let links = links
+        .into_iter()
+        .map(|(cid, len)| dag_pb::PbLink {
+            hash: Some(cid.to_bytes()),
+            // TODO: notes say that leaf chunks have "", does that mean that
+            // stem nodes should have None?
+            name: Some("".into()),
+            tsize: Some(len as u64),
+        })
+        .collect();
+
+    // PBNode.Data
+    let inner = unixfs_pb::Data {
+        r#type: DataType::File as i32,
+        ..Default::default()
+    };
+
+    // create PBNode
+    let outer = encode_unixfs_pb(&inner, links)?;
+
+    if is_root {
+        return Ok(UnixfsNode::Directory(Node { inner, outer }));
+    }
+
+    // create UnixfsNode
+    Ok(UnixfsNode::File(Node { inner, outer }))
+}
+
+// Leaf and Stem nodes are the two types of nodes that can exist in the tree
+// Leaf nodes encode to `UnixfsNode::Raw`
+// Stem nodes encode to `UnixfsNode::File`
+// Root nodes encode to `UnixfsNode::Directory`
```

Comment:
Okay, so in the original `file.encode()` the root node was a `UnixfsNode::Directory`. But we always talked about `UnixfsNode::File`s. Wanted to double check, what is the correct format for the root node?

I know that to preserve the file name we need to wrap the file in a directory, but that happens in `add_file`.